cmd/ddev/cmd/flags.go

```diff
@@ -0,0 +1,357 @@
+	}
+
+	return nil
+}
+
+func (v *typeValue) validate() error {
+	// Check type and set default if empty
+	if *v == "" {
+		*v = FtBool
+	}
+
+	// Check type is valid
+	implemented, found := ValidTypes[*v]
+
+	if !found {
+		return formatErrorItem(2, "-", "type '%s' is not known", *v)
+	} else if !implemented {
+		return formatErrorItem(2, "-", "type '%s' is not implemented", *v)
+	}
+
+	return nil
+}
+
+func (v *defValueValue) validate(typ typeValue) error {
+	if *v != "" {
+		return nil
+	}
+
+	// Init the DefValue
+	switch typ {
+	case FtBool:
+		*v = defValueValue(strconv.FormatBool(false))
+	case FtCount, FtDuration, FtFloat32, FtFloat64, FtInt, FtInt8, FtInt16, FtInt32, FtUint, FtUint8, FtUint16, FtUint32:
+		*v = defValueValue(strconv.FormatInt(0, 10))
+	case ftTest3: // used for testing only
+		*v = ""
+	default:
+		if implemented := ValidTypes[typ]; implemented {
+			// Mandatory implementation missing -> panic
+			panic(fmt.Sprintf("Error implementation of DefValue validation missing for type '%s'", typ))
+		}
+	}
+
+	return nil
+}
+
+func (v *noOptDefValValue) validate() error {
+	return nil
+}
+
+func (v *annotationsValue) validate() error {
+	return nil
+}
+
+func (f *Flag) validateFlag(longOptions *map[nameValue]bool, shortOptions *map[shorthandValue]nameValue) error {
+	errors := ""
+
+	// Chech all fields
+	errors += extractError(f.Name.validate(longOptions))
+	errors += extractError(f.Shorthand.validate(shortOptions, f.Name))
+	errors += extractError(f.Usage.validate())
+	errors += extractError(f.Type.validate())
+	errors += extractError(f.DefValue.validate(f.Type))
+	errors += extractError(f.NoOptDefVal.validate())
+	errors += extractError(f.Annotations.validate())
+
+	if errors != "" {
+		return fmt.Errorf("%s", errors)
+	}
+
+	return nil
+}
+
+// Init initializes the Flags structure.
+func (f *Flags) Init(commandName, script string) {
+	f.CommandName = commandName
+	f.Script = script
+}
+
+func (f *Flags) validateFlags(flags *FlagsDefinition) error {
+	var errors string
+
+	// Temporay vars to precheck for duplicated flags. It's still possible
+	// other commands will introduce the same flags which is tested
+	// afterwards by cobra.
+	long := map[nameValue]bool{"help": true}
+	short := map[shorthandValue]nameValue{"h": "help"}
+
+	for i := range *flags {
+		flag := &(*flags)[i]
+
+		// Additional validations of the flag fields
+		flagErrors := extractError(flag.validateFlag(&long, &short))
+
+		if flagErrors != "" {
+			errors += extractError(formatErrorItem(1, "*", "for flag '%s':%s", flag.Name, flagErrors))
+		}
+	}
+
+	if errors != "" {
+		return fmt.Errorf("The following problems were found in the flags definition of the command '%s' in '%s':%s", f.CommandName, f.Script, errors)
+	}
+
+	return nil
+}
+
+// LoadFromJSON imports the defs provided by the custom command as json into
+// the flags structure.
+func (f *Flags) LoadFromJSON(data string) error {
+	if data == "" {
+		return nil
+	}
+
+	var err error
```

Comment:
Explicitly defining err isn't necessary or particularly useful. Just use err := ... on the first error. OTOH, it doesn't actually do any harm.

Reply:
Yes, I'm aware of this. I did it like this because the var is used multiple times and for me it's better readable like this for this case.